bpo-43766: Implement PEP 647 (User-Defined Type Guards) in typing.py (#25282)
python/cpython: Doc/library/typing.rst

   .. versionadded:: 3.9

.. data:: TypeGuard

   Special typing form used to annotate the return type of a user-defined

>    type guard function.  ``TypeGuard`` only accepts a single type argument.

Reviewer: I'd add right here something like this:
```suggestion
   type guard function.  ``TypeGuard`` only accepts a single type argument.
   At runtime, functions marked this way return a boolean.
```
This would hopefully allay the repeatedly expressed worries that writing `-> TypeGuard[X]` obfuscates that the return type of such functions is `bool`.

(Maybe ``` (i.e., ``True`` or ``False``)``` is redundant.)

   ``TypeGuard`` aims to benefit *type narrowing* - a technique used by static
   type checkers to determine a more precise type of an expression within a
   program's code flow.  Usually type narrowing is done by analyzing
   conditional code flow and applying the narrowing to a block of code.  The

>    conditional expression here is sometimes referred to as a "type guard".

Reviewer: I'd add an example; either using `isinstance(x, C)` or `x is not None`, since those are the most commonly used (non-user-defined) type guards.

>    Sometimes, a type guard can be a complex checking function.

Reviewer: I'd phrase this without using "complex". Instead, I'd emphasize that sometimes instead of combining `isinstance()` and/or `is None` checks, it is simpler to have a user-defined function that acts as a type guard, esp. if the guard is either complex and used repeatedly, or not expressible using existing checks (e.g. if it needs to inspect the presence or value of attributes or dict items).

>    These type guard functions require ``TypeGuard`` to narrow their input types

Reviewer: I'd repeat a few times that we're talking about user-defined type guard functions here, not about the built-in functions such as `isinstance()` which are used to compose "built-in" type guards (not sure whether "built-in" is the right term actually).

   as the static type checker usually does not have enough information to
   statically infer them.

>    A ``TypeGuard`` tells the static type checker that for a given function:

Reviewer: ```suggestion
   Using ``-> TypeGuard`` tells the static type checker that for a given function:
```

>       1. The return value is a boolean.

Reviewer: (Fine to repeat it here, this section is more formal than the summary in the first paragraph.)

>       2. If the return value was "truthy", the type of the input to the

Reviewer: Let's not get into truthiness here -- let's just say ``` ``True`` ```. Also, s/was/is/.

Separately, I wonder if we need to somehow distinguish more carefully between what happens in the static checker and at runtime. The checker *makes the assumption* that a `True` result implies that the expression used as the first argument has the type given in the `TypeGuard[X]` annotation. It behooves the implementation of the type guard function to honor this assumption at runtime, but neither the type checker nor the runtime has the ability to verify that it is so -- if the function "lies", that is a runtime bug for which the checker cannot take responsibility.

Author: Do you mean that for the current PEP, things like:

```python

def is_fake_str_list(val: List[object]) -> TypeGuard[List[str]]:
    '''Liar!'''
    return True

def func1(val: List[object]):
    if is_str_list(val):
        # Type of ``val`` is narrowed to List[str]
        print(" ".join(val))
    else:
        # Type of ``val`` remains as List[object]
        print("Not a list of strings!")

# Kaboom!
func([1,2,3])
```

Fail at runtime instead of being caught at "compile" time?

If so, I'm not sure if there's an easy way to fix that. One way would be for the static type checker to verify ``is_fake_str_list`` recursively. But this would bring about a crop of other problems, weird edge cases, and added implementation complexity for them.

Here's some comments from the PEP that I agree with (though it's talking about why there isn't strict narrowing, not runtime vs static checkers' responsibilities):

> However, there are many ways a determined or uninformed developer can subvert type safety -- most commonly by using cast or Any. If a Python developer takes the time to learn about and implement user-defined type guards within their code, it is safe to assume that they are interested in type safety and will not write their type guard functions in a way that will undermine type safety or produce nonsensical results.

Reviewer: Yes, that's what I meant. And of course there's no easy fix for that. I just want the docs to not mislead the reader into thinking that that is somehow checked.

>          function is specified by the type inside ``TypeGuard``.

Reviewer: (Also, similar wording changes as I proposed in the docstring.)

      For example::

         def is_str_list(val: List[object]) -> TypeGuard[List[str]]:
             '''Determines whether all objects in the list are strings'''
             return all(isinstance(x, str) for x in val)

         def func1(val: List[object]):
             if is_str_list(val):
                 # Type of ``val`` is narrowed to List[str]
                 print(" ".join(val))
             else:
                 # Type of ``val`` remains as List[object]
                 print("Not a list of strings!")

   In short, the form ``def foo(arg: TypeA) -> TypeGuard[TypeB]: ...``,

>    Means that if ``foo(arg)`` returned true, then ``arg`` narrows from

Reviewer: ```suggestion
   means that if ``foo(arg)`` returns ``True``, then ``arg`` narrows from
```

>    ``TypeA`` to ``TypeB``.

Reviewer: I'd add something explaining that this "narrowing" could in practice be a "widening" in case `TypeA` is a subclass of `TypeB`. (I think there's an example in the PEP, as well as an explanation why the checker should not make an exception in this case.)

>    Return statements within a type guard function should return ``bool``
>    values.

Reviewer: ```suggestion
   A type guard function should return a ``bool`` value.
```

   .. versionadded:: 3.9

.. data:: TypeGuard

   Special typing form used to annotate the return type of a user-defined
   type guard function.  ``TypeGuard`` only accepts a single type argument.
   At runtime, functions marked this way should return a boolean.

   ``TypeGuard`` aims to benefit *type narrowing* - a technique used by static
   type checkers to determine a more precise type of an expression within a
   program's code flow.  Usually type narrowing is done by analyzing
   conditional code flow and applying the narrowing to a block of code.  The

>    conditional expression here is sometimes referred to as a "type guard"::

Reviewer: I'm not sure such an elaborate example is needed.

      def func(val: Optional[Union[str, float]]):
          # Non-"None" type guard
          if val is not None:
              # Type of val is narrowed to ``Union[str, float]``
              # "isinstance" type guard
              if isinstance(val, str):
                  # Type of val is narrowed to ``str``
                  ...
              else:
                  # Else, type of val is narrowed to ``float``.
                  ...
          else:
              # Type of val remains Optional[Union[str, float]]
              ...

>    Sometimes, a type guard uses a user-defined checking function instead of
>    ``isinstance`` or ``is None`` checks.  These user-defined type guard
>    functions require ``TypeGuard`` to narrow their input types as the static
>    type checker usually does not have enough information to statically infer
>    them.

Reviewer: ```suggestion
   Sometimes it would be convenient to use a user-defined boolean function
   as a type guard.  Such a function should use ``TypeGuard[...]`` as its return type
   to alert static type checkers to this intention.
```

   Using  ``-> TypeGuard`` tells the static type checker that for a given
   function:

   1. The return value is a boolean.
   2. If the return value is ``True``, the type of its argument
      is the type inside ``TypeGuard``.

      For example::

         def is_str_list(val: List[object]) -> TypeGuard[List[str]]:
             '''Determines whether all objects in the list are strings'''
             return all(isinstance(x, str) for x in val)

         def func1(val: List[object]):
             if is_str_list(val):
                 # Type of ``val`` is narrowed to List[str]
                 print(" ".join(val))
             else:
                 # Type of ``val`` remains as List[object]
                 print("Not a list of strings!")

   If ``is_str_list`` is a class or instance method, then the type in
   ``TypeGuard`` maps to the type of the second parameter after ``cls`` or
   ``self``.

   In short, the form ``def foo(arg: TypeA) -> TypeGuard[TypeB]: ...``,
   means that if ``foo(arg)`` returns ``True``, then ``arg`` narrows from
   ``TypeA`` to ``TypeB``.

   .. note::

>       Strict type narrowing is not enforced - ``TypeB`` need not be a narrower
>       form of ``TypeA`` (it can even be a wider form) and this may lead to
>       type-unsafe results.  The main reason is to allow for things like
>       narrowing ``List[object]`` to ``List[str]`` which would fail under strict
>       narrowing as ``List`` is invariant.  The responsibility of
>       writing type-safe type guards is left to the user.

Reviewer: ```suggestion
      ``TypeB`` need not be a narrower
      form of ``TypeA`` -- it can even be a wider form.
      The main reason is to allow for things like
      narrowing ``List[object]`` to ``List[str]`` even though the latter
      is not a subtype of the former, since ``List`` is invariant.  The responsibility of
      writing type-safe type guards is left to the user.
```

      type-unsafe results.  The main reason is to allow for things like
      narrowing ``List[object]`` to ``List[str]`` which would fail under strict
      narrowing as ``List`` is invariant.  The responsibility of
      writing type-safe type guards is left to the user.  Furthermore, even if
      the type guard function passes type checks, it may still fail at runtime.
      The type guard function may perform erroneous checks and return wrong
      booleans. Consequently, the type it promises in ``TypeGuard[TypeB]`` may

>       not hold.

Reviewer: Hm, this feels too verbose. How about just this:
```suggestion
      writing correct type-safe type guards is left to the user.
```

Author: I'll send that patch in soon. We can always merge it after your vacation.
